### PR TITLE
VIVI-11880 Prefer non-dash tracks, select tracks deterministically

### PR DESCRIPTION
**Pull request by saintmagician**
Two improvements that help with troubleshooting ytdl issues

1. Prefer url track over dash track. (Dash track requires us to create a manifest XML. You can't easily verify that this manifest XML is correct. A url track gives us a url that you can just punch into a web browser). 

2. Always select tracks deterministically. Vimeo videos will often have multiple tracks that seem identical. We pick the first one yt-dlp gives us, which seems to be random. This is annoying because it means different ytdl service nodes can end up selecting different tracks. 

Also added unit tests. Dunno what people think about this... might be a bit controversial 😅 If people are happy with the idea of unit tests like this, I'll add a buildkite step for this repo that runs the tests. 

These unit tests check helper functions that are hard to verify if you simply hit the ytdl service. For example, the ytdl service may seem like it's working, but how do you verify that it actually selected the correct track? If you update the `videoTrackSort` function, how many different videos do you have to check before you are confident that tracks are being sorted correctly?

### CODE REVIEW

File: tests/test.js
Comment (lucashasil) [inline tests/test.js:1]: Personally don't know how useful these tests are in practice but I think any tests good or bad are better than none.  I guess these might catch if someone breaks a function or something which I suppose could be possible

File: tests/test.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    it('tiebreaker on format_id' , function() {
+        const a = { format_id: 'aaa', height: 1080, acodec: 'opus', protocol: 'dash', tbr: 3000 };
+        const b = { format_id: 'bbb', height: 1080, acodec: 'opus', protocol: 'dash', tbr: 3000 };
+
+        assert.equal([a, b].sort(videoTrackSort)[0], a);
+        assert.equal([b, a].sort(videoTrackSort)[0], a);
+    });
+});
+
+describe('filterFormatCodecs' , function() {
```
Comment (lucashasil) [inline tests/test.js]: I think this is good

File: index.js [hunk 2]
```diff
@@ -272,6 +309,13 @@ function filterFormatCodecs(acodec, format_id, protocol, vcodec) {
...
+function filterFormatFps(format) {
+  const { fps, height } = format;
   return ((height >= 1080 && fps <= 30) || height < 1080);
 }
+
+module.exports._private_testing = {
+  generateDurationString,
+  videoTrackSort,
+  filterFormatCodecs
+}
```
Comment (lucashasil) [inline index.js]: Missing newline

File: package.json [hunk 3]
```diff
@@ -2,5 +2,11 @@
   "name": "ytdl-process",
   "version": "1.0.3",
   "description": "Shared module for our youtube-dl processing",
-  "main": "index.js"
+  "main": "index.js",
+  "devDependencies": {
+    "mocha": "^10.4.0"
```
Comment (lucashasil) [inline package.json]: Why Mocha and not Jest or Vitest tho?

File: index.js (hunk 2 above)
Comment (saintmagician) [reply index.js]: fixed

**Comment by saintmagician** [comment]
@lucashasil i've switched over to Jest. Also added the buildkite stuff and it seems to work and the tests are running.

**Review by lucashasil — APPROVED** [review]

**Pull request merged**